source/extensions/common/wasm/exports.cc

```diff
@@ -0,0 +1,220 @@
+#include "extensions/common/wasm/wasm.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace Wasm {
+namespace Exports {
+
+// Any currently executing Wasm call context.
+#define WASM_CONTEXT(_c)                                                                           \
+  (ContextOrEffectiveContext(static_cast<Context*>((void)_c, current_context_)))
+// The id of the context which should be used for calls out of the VM in place of current_context_
+// above.
+
+namespace {
+
+inline Word wasmResultToWord(WasmResult r) { return Word(static_cast<uint64_t>(r)); }
+
+Context* ContextOrEffectiveContext(Context* context) {
+  if (effective_context_id_ == 0) {
+    return context;
+  }
+  auto effective_context = context->wasm()->getContext(effective_context_id_);
+  if (effective_context) {
+    return effective_context;
+  }
+  // The effective_context_id_ no longer exists, revert to the true context.
+  return context;
+}
+
+} // namespace
+
+Word get_configuration(void* raw_context, Word value_ptr_ptr, Word value_size_ptr) {
+  auto context = WASM_CONTEXT(raw_context);
+  auto value = context->getConfiguration();
+  if (!context->wasm()->copyToPointerSize(value, value_ptr_ptr.u64_, value_size_ptr.u64_)) {
+    return wasmResultToWord(WasmResult::InvalidMemoryAccess);
+  }
+  return wasmResultToWord(WasmResult::Ok);
+}
+
+Word set_effective_context(void* raw_context, Word context_id) {
```

Comment:
Calls out of the VM are done w.r.t a Context, e.g. stream or root context.  gRPC calls are passed to a RootContext since they can outlive a give stream (e.g. reporting events to an external system).  If the gRPC is (for example) filling a cache entry where there are multiple streams waiting on that cache entry then we need to be able to set the effective context to each of those streams so that they can be restarted and handled w.r.t. the new cached data.

However, I will pull this out and bring it back in a future PR which includes gRPC support.